File: Lab9ZgurskyiKI305/House.py
House.py
import datetime

class House:
    def __init__(self, address="No information", number_of_floors=0, has_garden=False):
        self.address = address
        self.number_of_floors = number_of_floors
        self.has_garden = has_garden
        self.writer = None

    def open_log_file(self):
        try:
            self.writer = open("log.txt", "a")
        except IOError as e:
            print(f"Помилка при відкритті файлу для запису: {e}")

    def close_log_file(self):
        try:
            if self.writer:
                self.writer.close()
        except IOError as e:
            print(f"Помилка при закритті файлу: {e}")

    def display_details(self):
        print(f"Адреса будинку: {self.address}")
        print(f"Кількість поверхів: {self.number_of_floors}")
        print(f"Наявність саду: {'Так' if self.has_garden else 'Ні'}")

    def log_message(self, message):
        try:
            if self.writer:
                timestamp = datetime.datetime.now()
                self.writer.write(f"[{timestamp}] {message} - {self}\n")
        except IOError as e:
            print(f"Помилка при записі до файлу: {e}")

    def set_number_of_floors(self, number_of_floors):
        self.number_of_floors = number_of_floors
        self.log_message(f"Встановлено кількість поверхів: {number_of_floors}")

    def set_address(self, address):
        self.address = address
        self.log_message(f"Оновлено адресу: {address}")

    def set_has_garden(self, has_garden):
        self.has_garden = has_garden
        self.log_message("Оновлено інформацію про сад.")

    def add_floor(self):
        self.number_of_floors += 1
        self.log_message("Додано поверх.")

    def remove_floor(self):
        if self.number_of_floors > 0:
            self.number_of_floors -= 1
            self.log_message("Видалено поверх.")
        else:
            self.log_message("Не можна видалити поверх. Кількість поверхів вже мінімальна.")

    def get_number_of_floors(self):
        self.log_message("Дана інформація про кількість поверхів.")
        return self.number_of_floors

    def get_address(self):
        return self.address

    def has_garden(self):
        self.log_message("Дана інформація про наявність саду.")
        return self.has_garden















HouseApp.py:
from house import House

class HouseApp:
    def main(self):
        # Create a house
        house = House("123 Main St.", 2, True)

        # Display house details
        house.display_details()

        # Update house details
        house.set_number_of_floors(3)
        house.set_address("456 Elm St.")
        house.set_has_garden(False)

        # Display updated house details
        house.display_details()

        house.to_string()

if __name__ == "__main__":
    app = HouseApp()
    app.main()









OfficeCenter.py:
from house import House
import datetime

class OfficeCenter(House):
    def __init__(self, address, number_of_floors, office_space, has_meeting_room, number_of_desks, has_projector, has_whiteboard):
        super().__init__(address, number_of_floors)
        self.office_space = office_space
        self.has_meeting_room = has_meeting_room
        self.number_of_desks = number_of_desks
        self.has_projector = has_projector
        self.has_whiteboard = has_whiteboard

    def get_office_space(self):
        return self.office_space

    def set_office_space(self, office_space):
        self.office_space = office_space

    def allocate_office_space(self, square_meters):
        self.office_space += square_meters

    def equip_meeting_room(self, projector, whiteboard):
        self.has_meeting_room = True
        self.has_projector = projector
        self.has_whiteboard = whiteboard

    def add_desks(self, desks_to_add):
        self.number_of_desks += desks_to_add

    def remove_desks(self, desks_to_remove):
        if desks_to_remove > self.number_of_desks:
            print("Cannot remove more desks than available.")
        else:
            self.number_of_desks -= desks_to_remove

    def add_projector(self, has_projector):
        self.has_projector = has_projector

    def add_whiteboard(self, has_whiteboard):
        self.has_whiteboard = has_whiteboard

    def add_projector2(self, has_projector):
        print("Projector was added")
        has_projector = True

    def write_address_to_file(self):
        address = self.get_address()

        try:
            with open("Address.txt", "w") as writer:
                writer.write(address)
        except IOError as e:
            print(f"Помилка при записі у файл: {e}")

        return address

    def __str__(self):
        return self.write_address_to_file()










OfficeInterface.py:
class OfficeInterface:
    def get_office_space(self):
        pass

    def set_office_space(self, office_space):
        pass

    def allocate_office_space(self, square_meters):
        pass

    def equip_meeting_room(self, projector, whiteboard):
        pass

    def add_desks(self, desks_to_add):
        pass

    def remove_desks(self, desks_to_remove):
        pass

    def add_projector(self, has_projector):
        pass

    def add_whiteboard(self, has_whiteboard):
        pass

    def write_address_to_file(self):
        pass









OfficeInterface2.py:
class OfficeInterface2:
    def add_projector2(self, has_projector):
        pass

    def add_whiteboard2(self, has_whiteboard):
        pass

File: Lab2ZgurskiyKI305/src/main/java/org/example/House.java
package org.example;

import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;

/**
 * The <code>House</code> class represents a house and its operations.
 * It includes functionality for managing the number of floors, addresses,
 * gardens, and provides information about the house.
 * This class also logs events to a file named "Log.txt".
 *
 * @version 1.0
 */
public class House {
    private FileWriter writer; // Field to store a reference to the file writing stream

    private String address;
    private int numberOfFloors;
    private boolean hasGarden;

    // Static variables for tracking parameters
    public static int param3 = 0;
    public static int param2 = 0;

    private static int countOfHousesWithAddress = 0;

    /**
     * Gets the count of houses with the address "Вулиця Лінкольна".
     *
     * @return The count of houses with the specified address.
     */
    public static int getCountOfHousesWithAddress() {
        return countOfHousesWithAddress;
    }
    /**
     * Default constructor for the house.
     */
    public House() {
        address = "No information";
        numberOfFloors = 0;
        hasGarden = false;
    }

    /**
     * Parameterized constructor for the house.
     *
     * @param address      Specifies initial address.
     * @param numberOfFloors The initial number of floors.
     * @param hasGarden    Specifies if the garden initially exists.
     */
    public House(String address, int numberOfFloors, boolean hasGarden) {
        this.address = address;
        this.numberOfFloors = numberOfFloors;
        this.hasGarden = hasGarden;
        if ("Вулиця Лінкольна".equals(address)) {
            countOfHousesWithAddress++;
        }
    }

    /**
     * Constructs a house with the specified address and number of floors, defaulting to no garden.
     *
     * @param address      The address of the house.
     * @param numberOfFloors The number of floors in the house.
     */
    public House(String address, int numberOfFloors) {
        this(address, numberOfFloors, false);

    }

    /**
     * Constructs a house with the specified address and defaults to one floor and no garden.
     *
     * @param address The address of the house.
     */
    public House(String address) {
        this(address, 1);

    }

    // Methods

    /**
     * Open the file for writing.
     */
    public void openLogFile() {
        try {
            writer = new FileWriter("log.txt", true);
        } catch (IOException e) {
            System.err.println("Error opening the file for writing: " + e.getMessage());
        }
    }

    /**
     * Close the file after ending writing.
     */
    public void closeLogFile() {
        try {
            if (writer != null) {
                writer.close();
            }
        } catch (IOException e) {
            System.err.println("Error closing the file: " + e.getMessage());
        }
    }

    /**
     * Display details of the house.
     */
    public void displayDetails() {
        System.out.println("House address: " + address);
        System.out.println("Number of floors: " + numberOfFloors);
        System.out.println("Has a garden: " + (hasGarden ? "Yes" : "No"));
    }

    /**
     * Log a message to the file.
     *
     * @param message The message to log.
     */
    private void logMessage(String message) {
        try (FileWriter writer = new FileWriter("log.txt", true)) {
            LocalDateTime timestamp = LocalDateTime.now();
            writer.write("[" + timestamp + "] " + message + " - " + this + "\n");
        } catch (IOException e) {
            System.err.println("Error writing to the file: " + e.getMessage());
        }
    }

    /**
     * Set the number of floors for the house.
     *
     * @param numberOfFloors The number of floors to set.
     */
    public void setNumberOfFloors(int numberOfFloors) {
        this.numberOfFloors = numberOfFloors;
        logMessage("Set the number of floors to: " + numberOfFloors);
    }

    /**
     * Set the address for the house.
     *
     * @param address The address to set.
     */
    public void setAddress(String address) {
        this.address = address;
        logMessage("Updated address: " + address);
    }

    /**
     * Set whether the house has a garden or not.
     *
     * @param hasGarden True if the house has a garden, false otherwise.
     */
    public void setHasGarden(boolean hasGarden) {
        this.hasGarden = hasGarden;
        logMessage("Updated garden information.");
    }

    /**
     * Add a floor to the house.
     */
    public void addFloor() {
        numberOfFloors++;
        logMessage("Added a floor.");
    }

    /**
     * Remove a floor from the house.
     * If the number of floors is already at the minimum, log a message accordingly.
     */
    public void removeFloor() {
        if (numberOfFloors > 0) {
            numberOfFloors--;
            logMessage("Removed a floor.");
        } else {
            logMessage("Cannot remove a floor. The number of floors is already at the minimum.");
        }
    }

    /**
     * Get the number of floors for the house.
     *
     * @return The number of floors.
     */
    public int getNumberOfFloors() {
        logMessage("Retrieved information about the number of floors.");
        return numberOfFloors;
    }

    /**
     * Get the address of the house.
     *
     * @return The address.
     */
    public String getAddress() {
        logMessage("Retrieved information about the address.");
        return address;
    }

    /**
     * Check if the house has a garden.
     *
     * @return True if the house has a garden, false otherwise.
     */
    public boolean hasGarden() {
        logMessage("Retrieved information about the garden.");
        return hasGarden;
    }

    // Additional methods

}
